keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,169 @@
+
+In the remaining of this document:
+1. *Job* refers to any kind job supported by Kueue including `batch/job`, `MPIJob`, `RayJob`, etc.
+1. *Pods of a Workload* means the same as *Pods of a Job* and vice-versa.
+
+
+## Motivation
+
+Currently, the assigned to a Job are reclaimed by Kueue only when the whole Job finishes.
+For Jobs with multiple Pods, the resources will be reclaimed only after the last Pod of the Job finishes.
+This is not efficient as the Jobs might have different needs during execution, for instance the needs of a `batch/job` having the `parallelism` equal to `completions` will decrease with every Pod finishing it's execution.
+
+### Goals
+
+- Utilize the unused resources of the successfully completed Pods of a running Job.
+
+### Non-Goals
+
+- Preempting a Workload or Pods of a Workload to free resources for a pending Workload.
+- Partially admitting a Workload.
+- Monitor the Job's pod execution.
+
+## Proposal
+
+Reclaiming the resources of the succeeded Pods of a running Job as soon as the Pod completes its execution.
+
+We propose to add a new field `.status.ReclaimablePods` to the Workload API. `.status.ReclaimablePods` is a list that holds the count of Pods belonging to a PodSet whose resources are no longer needed and could be reclaimed by Kueue.
+
+### Pod Completion Accounting
+
+Since the ability of actively monitoring pods execution of a job is not in the scope of the core Kueue implementation, the number of pods for which the resources are no longer needed for each PodSet should be reported by each framework specific `GenericJob` implementation.
+
+For this purpose the `GenericJob` interface should be changed to either:
+
+1. Add an additional method able to report this
+
+```go
+type GenericJob interface {
+    // ...
+
+   // Get reclaimable pods.
+   ReclaimablePods() map[string]int
+}
+
+```
+
+2. Modify the current `Finished` method
+
+```go
+type GenericJob interface {
+   // ...
+
+   // Get reclaimable pods.
+   Finished() (condition metav1.Condition, reclaimablePods map[string]int)
+}
+
+```
+
+#### Example (`batch/job`)
+
+Having a job defined with **P** `parallelism` and **C** `completions`, and **n** number of completed pod executions,
+the expected result should be:
+
+```go
+map[string]int{
+    "main": P - max(P,(C-n))
+}
+```
+### API
+
+A new field `ReclaimablePods` is added to the `.status` of Workload API.
+
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+
+    // ...
+
+    // list of count of Pods of a PodSet with reclaimable resources 
+    // +optional
+    ReclaimablePods []ReclaimablePod `json:"reclaimablePod"`
```

Comment:
this would fail in the exceptions for using `map[string]int32`, given that there is no world where we would need another member of the map, is there?
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Reply:
We wold go with slice and
- if a pod set is not mentioned , the reclaimable cont is considered 0
- if an entry has a name not found in the `spec.podSets` is going to be ignored